File: src/emby_range_cache_proxy/prefetch.py
from __future__ import annotations

import asyncio
import logging
from dataclasses import dataclass
from typing import Awaitable, Callable

from aiohttp import ClientError

from .cache import adaptive_head_tail, cache_key
from .config import Config, MiddleCacheConfig, PrefetchConfig
from .middle_cache import MiddleRangeCache
from .models import ByteRange, MediaSource
from .origin import OriginClient, OriginError
from .state import PlaybackSessionRecord, PrefetchTaskRecord, SessionStateStore

LOGGER = logging.getLogger(__name__)


@dataclass
class PrefetchRunResult:
    completed: int = 0
    failed: int = 0
    skipped: int = 0


class PrefetchSourceMismatch(Exception):
    pass


def short_hash(value: str | None) -> str:
    return "none" if value is None else value[:12]


class BandwidthLimiter:
    def __init__(
        self,
        *,
        bytes_per_second: int,
        sleep: Callable[[float], Awaitable[None]] = asyncio.sleep,
    ) -> None:
        if bytes_per_second <= 0:
            raise ValueError("bytes_per_second must be positive")
        self.bytes_per_second = bytes_per_second
        self.sleep = sleep

    async def consume(self, byte_count: int) -> None:
        if byte_count <= 0:
            return
        await self.sleep(byte_count / self.bytes_per_second)


class PrefetchWorker:
    def __init__(
        self,
        config: Config,
        store: SessionStateStore,
        middle_cache: MiddleRangeCache,
        *,
        source_lookup: dict[tuple[str, str], str] | None = None,
    ) -> None:
        self.config = config
        self.store = store
        self.middle_cache = middle_cache
        self.source_lookup = dict(source_lookup or {})
        self.limiter = BandwidthLimiter(
            bytes_per_second=config.prefetch.bandwidth_bytes_per_second
        )

    async def run_once(self, *, now: float) -> PrefetchRunResult:
        if not self.config.prefetch.enabled or not self.config.middle_cache.enabled:
            return PrefetchRunResult()

        result = PrefetchRunResult()
        for _ in range(self.config.prefetch.concurrency):
            tasks = await asyncio.to_thread(
                self.store.claim_prefetch_tasks,
                limit=1,
                now=now,
                running_stale_seconds=self.config.prefetch.error_backoff_seconds,
            )
            if not tasks:
                break
            task = tasks[0]
            task_result = await self._run_task(task, now=now)
            result.completed += task_result.completed
            result.failed += task_result.failed
            result.skipped += task_result.skipped
        return result

    async def _run_task(
        self, task: PrefetchTaskRecord, *, now: float
    ) -> PrefetchRunResult:
        _log_prefetch_event("prefetch_started", task)
        byte_range = ByteRange(task.start, task.end)
        if byte_range.length > self.middle_cache.max_bytes:
            await asyncio.to_thread(
                self.store.skip_prefetch_task,
                task.id,
                error_class="RangeTooLarge",
                now=now,
                expected_attempts=task.attempts,
            )
            _log_prefetch_event("prefetch_skipped", task, reason="RangeTooLarge")
            return PrefetchRunResult(skipped=1)

        url = self.source_lookup.get((task.item_id, task.media_source_id))
        if url is None:
            source_metadata = await asyncio.to_thread(
                self.store.get_source_metadata,
                task.item_id,
                task.media_source_id,
                task.cache_key,
            )
            if source_metadata is not None:
                url = source_metadata.origin_url
        if url is None:
            await asyncio.to_thread(
                self.store.skip_prefetch_task,
                task.id,
                error_class="SourceUnavailable",
                now=now,
                retry_after_seconds=self.config.prefetch.error_backoff_seconds,
                expected_attempts=task.attempts,
            )
            _log_prefetch_event(
                "prefetch_skipped", task, reason="SourceUnavailable"
            )
            return PrefetchRunResult(skipped=1)

        try:
            data = bytearray()
            async with OriginClient(
                chunk_bytes=self.config.cache.chunk_bytes
            ) as origin:
                metadata = await origin.head(url)
                expected_key = cache_key(
                    MediaSource(
                        item_id=task.item_id,
                        media_source_id=task.media_source_id,
                        path=url,
                        protocol="Http",
                        size=metadata.size,
                    ),
                    metadata,
                )
                if expected_key != task.cache_key:
                    raise PrefetchSourceMismatch("prefetch source metadata mismatch")
                async with origin.open_range(
                    url,
                    byte_range,
                    size=metadata.size,
                ) as response:
                    async for chunk in response.content.iter_chunked(
                        self.config.cache.chunk_bytes
                    ):
                        if not chunk:
                            continue
                        await self.limiter.consume(len(chunk))
                        data.extend(chunk)

            stored = await asyncio.to_thread(
                self._store_completed_task,
                task,
                byte_range,
                bytes(data),
                now,
            )
            if not stored:
                _log_prefetch_event("prefetch_skipped", task, reason="PublishRace")
                return PrefetchRunResult(skipped=1)
            _log_prefetch_event("prefetch_complete", task)
            return PrefetchRunResult(completed=1)
        except asyncio.CancelledError:
            await asyncio.to_thread(
                self.store.requeue_prefetch_task,
                task.id,
                now=now,
                error_class="CancelledError",
                expected_attempts=task.attempts,
            )
            raise
        except (
            PrefetchSourceMismatch,
            OriginError,
            ClientError,
            asyncio.TimeoutError,
            TimeoutError,
            OSError,
            ValueError,
        ) as error:
            retry_after_seconds = (
                None
                if isinstance(error, PrefetchSourceMismatch)
                else self.config.prefetch.error_backoff_seconds
            )
            await asyncio.to_thread(
                self.store.fail_prefetch_task,
                task.id,
                error_class=error.__class__.__name__,
                now=now,
                retry_after_seconds=retry_after_seconds,
                expected_attempts=task.attempts,
            )
            _log_prefetch_event(
                "prefetch_failed", task, error_class=error.__class__.__name__
            )
            return PrefetchRunResult(failed=1)

    def _store_completed_task(
        self,
        task: PrefetchTaskRecord,
        byte_range: ByteRange,
        data: bytes,
        now: float,
    ) -> bool:
        stored = self.middle_cache.store_prefetch_block(
            task.id,
            expected_attempts=task.attempts,
            key=task.cache_key,
            byte_range=byte_range,
            data=data,
            now=now,
        )
        if not stored:
            return False
        self.middle_cache.evict_expired(now=now)
        self.middle_cache.evict_lru_if_needed()
        return True


def _log_prefetch_event(
    event: str,
    task: PrefetchTaskRecord,
    *,
    reason: str | None = None,
    error_class: str | None = None,
) -> None:
    LOGGER.info(
        "%s item_id=%s media_source_id=%s cache_key=%s range=%s-%s "
        "reason=%s error_class=%s",
        event,
        task.item_id,
        task.media_source_id,
        short_hash(task.cache_key),
        task.start,
        task.end,
        reason or "none",
        error_class or "none",
    )


def align_down(value: int, alignment: int) -> int:
    return value - (value % alignment)


def align_up(value: int, alignment: int) -> int:
    remainder = value % alignment
    if remainder == 0:
        return value
    return value + alignment - remainder


def middle_prefetch_window_bytes(head_size: int, prefetch: PrefetchConfig) -> int:
    if head_size <= 0:
        return 0
    return min(head_size, prefetch.window_bytes, prefetch.max_session_bytes)


def middle_prefetch_overlap_bytes(
    window_bytes: int, prefetch: PrefetchConfig
) -> int:
    if window_bytes <= 1:
        return 0
    return min(prefetch.resume_overlap_bytes, window_bytes // 2)


def plan_middle_ranges(
    *,
    media_size: int,
    head_size: int,
    tail_size: int,
    anchor_offset: int,
    queued_until: int | None,
    prefetch: PrefetchConfig,
    middle_cache: MiddleCacheConfig,
) -> list[ByteRange]:
    segment = middle_cache.segment_bytes
    window_bytes = middle_prefetch_window_bytes(head_size, prefetch)
    if window_bytes <= 0:
        return []
    head_end = min(head_size, media_size) - 1
    tail_start = max(0, media_size - tail_size)
    middle_start = head_end + 1
    middle_end = tail_start - 1
    if middle_start > middle_end:
        return []

    if queued_until is None:
        if anchor_offset <= head_end:
            return []
        overlap_bytes = middle_prefetch_overlap_bytes(window_bytes, prefetch)
        start = max(middle_start, anchor_offset - overlap_bytes)
        if window_bytes > segment:
            start = max(middle_start, align_down(start, segment))
    else:
        start = max(middle_start, queued_until + 1)
    session_end = min(start + window_bytes - 1, middle_end)

    ranges: list[ByteRange] = []
    current = start
    while current <= session_end:
        segment_end = current + segment - 1
        if current % segment:
            segment_end = align_up(current, segment) - 1
        end = min(segment_end, session_end, middle_end)
        if end >= middle_start and current <= middle_end:
            ranges.append(ByteRange(max(current, middle_start), end))
        current = end + 1
    return ranges


def enqueue_prefetch_for_session(
    store: SessionStateStore,
    session: PlaybackSessionRecord,
    *,
    prefetch: PrefetchConfig,
    middle_cache: MiddleCacheConfig,
    now: float,
    priority: int,
) -> int:
    head_size, tail_size = adaptive_head_tail(session.media_size)
    target_ranges = plan_middle_ranges(
        media_size=session.media_size,
        head_size=head_size,
        tail_size=tail_size,
        anchor_offset=session.last_range_end,
        queued_until=None,
        prefetch=prefetch,
        middle_cache=middle_cache,
    )
    if not target_ranges:
        return 0
    target_start = target_ranges[0].start
    target_end = target_ranges[-1].end
    queued_until = (
        session.queued_until
        if session.queued_until is not None
        and target_start <= session.queued_until < target_end
        else None
    )

    ranges = plan_middle_ranges(
        media_size=session.media_size,
        head_size=head_size,
        tail_size=tail_size,
        anchor_offset=session.last_range_end,
        queued_until=queued_until,
        prefetch=prefetch,
        middle_cache=middle_cache,
    )

    inserted = 0
    highest_end: int | None = None
    stop_queueing = False
    for byte_range in ranges:
        if stop_queueing:
            break
        if byte_range.start > target_end:
            break
        if byte_range.end > target_end:
            byte_range = ByteRange(byte_range.start, target_end)
        missing_ranges = _subtract_ranges(
            byte_range,
            store.reusable_prefetch_ranges(session.cache_key, byte_range),
        )
        if not missing_ranges:
            highest_end = byte_range.end
            continue

        for missing_range in missing_ranges:
            task = store.enqueue_prefetch_task(
                session.item_id,
                session.media_source_id,
                session.cache_key,
                missing_range.start,
                missing_range.end,
                priority=priority,
                now=now,
                max_queue_depth=prefetch.max_queue_depth,
            )
            if task is None:
                if store.prefetch_task_exists(
                    session.cache_key,
                    missing_range.start,
                    missing_range.end,
                ):
                    highest_end = missing_range.end
                    continue
                stop_queueing = True
                break
            inserted += 1
            LOGGER.info(
                "prefetch_queued item_id=%s media_source_id=%s session=%s "
                "cache_key=%s range=%s-%s priority=%s",
                session.item_id,
                session.media_source_id,
                short_hash(session.session_hash),
                short_hash(session.cache_key),
                missing_range.start,
                missing_range.end,
                priority,
            )
            highest_end = missing_range.end
        if stop_queueing:
            break

    if highest_end is not None:
        store.update_session_queued_until(
            session.session_hash,
            highest_end,
            now=now,
        )
    return inserted


def _subtract_ranges(
    byte_range: ByteRange, existing_ranges: list[ByteRange]
) -> list[ByteRange]:
    missing: list[ByteRange] = []
    current = byte_range.start
    for existing in existing_ranges:
        if existing.end < current:
            continue
        if existing.start > byte_range.end:
            break
        if existing.start > current:
            missing.append(ByteRange(current, min(existing.start - 1, byte_range.end)))
        current = max(current, existing.end + 1)
        if current > byte_range.end:
            break
    if current <= byte_range.end:
        missing.append(ByteRange(current, byte_range.end))
    return missing

File: go/internal/prefetch/prefetch.go
package prefetch

import (
	"fmt"
	"io"
	"net/url"
	"sync"
	"time"

	"github.com/xmm2022/emby-range-cache-proxy/go/internal/cache"
	"github.com/xmm2022/emby-range-cache-proxy/go/internal/config"
	"github.com/xmm2022/emby-range-cache-proxy/go/internal/middle"
	"github.com/xmm2022/emby-range-cache-proxy/go/internal/model"
	"github.com/xmm2022/emby-range-cache-proxy/go/internal/origin"
	"github.com/xmm2022/emby-range-cache-proxy/go/internal/ranges"
	"github.com/xmm2022/emby-range-cache-proxy/go/internal/state"
)

type RunResult struct {
	Completed int
	Failed    int
	Skipped   int
}

type Worker struct {
	Prefetch    config.PrefetchConfig
	CacheConfig config.CacheConfig
	Store       *state.Store
	Middle      *middle.Cache
	Origin      *origin.Client
	RunningHook func(delta int)
	limiter     *BandwidthLimiter
	originLocks *originLimiter
}

func NewWorker(prefetch config.PrefetchConfig, cacheConfig config.CacheConfig, store *state.Store, middleCache *middle.Cache) *Worker {
	limiter := NewBandwidthLimiter(prefetch.BandwidthBytesPerSecond)
	if prefetch.BandwidthBytesPerSecond <= 0 {
		limiter = NewBandwidthLimiter(1)
	}
	return &Worker{
		Prefetch:    prefetch,
		CacheConfig: cacheConfig,
		Store:       store,
		Middle:      middleCache,
		Origin:      origin.NewClient(cacheConfig.ChunkBytes),
		limiter:     limiter,
		originLocks: newOriginLimiter(prefetch.PerOriginConcurrency),
	}
}

func (w *Worker) RunOnce(now float64) (RunResult, error) {
	if !w.Prefetch.Enabled {
		return RunResult{}, nil
	}
	concurrency := w.Prefetch.Concurrency
	if concurrency <= 0 {
		concurrency = 1
	}
	tasks, err := w.Store.ClaimPrefetchTasks(concurrency, now, w.Prefetch.ErrorBackoffSeconds)
	if err != nil {
		return RunResult{}, err
	}
	if len(tasks) == 0 {
		return RunResult{}, nil
	}
	if w.RunningHook != nil {
		w.RunningHook(len(tasks))
		defer w.RunningHook(-len(tasks))
	}
	var result RunResult
	var mu sync.Mutex
	var wg sync.WaitGroup
	for _, task := range tasks {
		wg.Add(1)
		go func(task state.PrefetchTaskRecord) {
			defer wg.Done()
			taskResult := w.runTask(task, now)
			mu.Lock()
			result.Completed += taskResult.Completed
			result.Failed += taskResult.Failed
			result.Skipped += taskResult.Skipped
			mu.Unlock()
		}(task)
	}
	wg.Wait()
	return result, nil
}

func (w *Worker) runTask(task state.PrefetchTaskRecord, now float64) RunResult {
	byteRange := model.ByteRange{Start: task.Start, End: task.End}
	if byteRange.Length() > w.Middle.MaxBytes {
		_ = w.Store.SkipPrefetchTask(task.ID, "RangeTooLarge", now, 0, task.Attempts)
		return RunResult{Skipped: 1}
	}
	sourceMeta, err := w.Store.GetSourceMetadata(task.ItemID, task.MediaSourceID, task.CacheKey)
	if err != nil || sourceMeta == nil {
		_ = w.Store.SkipPrefetchTask(task.ID, "SourceUnavailable", now, w.Prefetch.ErrorBackoffSeconds, task.Attempts)
		return RunResult{Skipped: 1}
	}
	release := w.originLocks.acquire(sourceMeta.OriginURL)
	defer release()
	meta, err := w.Origin.Head(sourceMeta.OriginURL)
	if err != nil {
		_ = w.Store.FailPrefetchTask(task.ID, "OriginError", now, w.Prefetch.ErrorBackoffSeconds, task.Attempts)
		return RunResult{Failed: 1}
	}
	expectedKey := cache.Key(model.MediaSource{
		ItemID:        task.ItemID,
		MediaSourceID: task.MediaSourceID,
		Path:          sourceMeta.OriginURL,
		Protocol:      "Http",
	}, meta)
	if expectedKey != task.CacheKey {
		_ = w.Store.FailPrefetchTask(task.ID, "PrefetchSourceMismatch", now, 0, task.Attempts)
		return RunResult{Failed: 1}
	}
	body, err := w.Origin.OpenRange(meta.URL, byteRange, meta.Size)
	if err != nil {
		_ = w.Store.FailPrefetchTask(task.ID, "OriginError", now, w.Prefetch.ErrorBackoffSeconds, task.Attempts)
		return RunResult{Failed: 1}
	}
	defer body.Close()
	reader := io.Reader(body)
	if w.limiter != nil {
		reader = &limitedReader{reader: body, limiter: w.limiter}
	}
	ok, err := w.Middle.StorePrefetchBlockFromReader(task.ID, task.Attempts, task.CacheKey, byteRange, reader, now)
	if err != nil {
		_ = w.Store.FailPrefetchTask(task.ID, "StoreError", now, w.Prefetch.ErrorBackoffSeconds, task.Attempts)
		return RunResult{Failed: 1}
	}
	if !ok {
		return RunResult{Skipped: 1}
	}
	_, _ = w.Middle.EvictExpired(now)
	_, _ = w.Middle.EvictLRUIfNeeded()
	return RunResult{Completed: 1}
}

func PlanMiddleRanges(mediaSize, headSize, tailSize, anchorOffset int64, queuedUntil *int64, prefetch config.PrefetchConfig, middleCache config.MiddleCacheConfig) []model.ByteRange {
	segment := middleCache.SegmentBytes
	if segment <= 0 {
		return nil
	}
	headEnd := headSize - 1
	if headEnd >= mediaSize {
		headEnd = mediaSize - 1
	}
	tailStart := mediaSize - tailSize
	if tailStart < 0 {
		tailStart = 0
	}
	middleStart := headEnd + 1
	middleEnd := tailStart - 1
	if middleStart > middleEnd {
		return nil
	}
	start, sessionEnd, ok := planMiddleWindow(middleStart, middleEnd, segment, headEnd, anchorOffset, queuedUntil, prefetch)
	if !ok {
		return nil
	}
	if start < middleStart {
		start = middleStart
	}
	if sessionEnd > middleEnd {
		sessionEnd = middleEnd
	}
	if start > sessionEnd {
		return nil
	}
	var out []model.ByteRange
	current := start
	for current <= sessionEnd {
		segmentEnd := current + segment - 1
		if current%segment != 0 {
			segmentEnd = alignUp(current, segment) - 1
		}
		end := segmentEnd
		if end > sessionEnd {
			end = sessionEnd
		}
		if end > middleEnd {
			end = middleEnd
		}
		if end >= middleStart && current <= middleEnd {
			startRange := current
			if startRange < middleStart {
				startRange = middleStart
			}
			out = append(out, model.ByteRange{Start: startRange, End: end})
		}
		current = end + 1
	}
	return out
}

func planMiddleWindow(middleStart, middleEnd, segment, headEnd, anchorOffset int64, queuedUntil *int64, prefetch config.PrefetchConfig) (int64, int64, bool) {
	if prefetch.ResumeBackBlocks > 0 || prefetch.ResumeForwardBlocks > 0 {
		if anchorOffset <= headEnd {
			return 0, 0, false
		}
		anchorBlockStart := alignDown(anchorOffset, segment)
		start := anchorBlockStart - int64(prefetch.ResumeBackBlocks)*segment
		sessionEnd := anchorBlockStart + int64(prefetch.ResumeForwardBlocks+1)*segment - 1
		if queuedUntil != nil {
			start = *queuedUntil + 1
		}
		return start, sessionEnd, true
	}

	windowBytes := minPositive(headEnd+1, prefetch.WindowBytes, prefetch.MaxSessionBytes)
	if windowBytes <= 0 {
		return 0, 0, false
	}
	var start int64
	if queuedUntil == nil {
		if anchorOffset <= headEnd {
			return 0, 0, false
		}
		overlap := prefetch.ResumeOverlapBytes
		if half := windowBytes / 2; overlap > half {
			overlap = half
		}
		start = anchorOffset - overlap
		if start < middleStart {
			start = middleStart
		}
		if windowBytes > segment {
			start = alignDown(start, segment)
			if start < middleStart {
				start = middleStart
			}
		}
	} else {
		start = *queuedUntil + 1
		if start < middleStart {
			start = middleStart
		}
	}
	sessionEnd := start + windowBytes - 1
	if sessionEnd > middleEnd {
		sessionEnd = middleEnd
	}
	return start, sessionEnd, true
}

func EnqueueForSession(store *state.Store, session state.PlaybackSessionRecord, prefetch config.PrefetchConfig, middleCache config.MiddleCacheConfig, now float64, priority int) (int, error) {
	headSize, tailSize := ranges.AdaptiveHeadTail(session.MediaSize)
	rangesToQueue := PlanMiddleRanges(session.MediaSize, headSize, tailSize, session.LastRangeEnd, nil, prefetch, middleCache)
	if len(rangesToQueue) == 0 {
		return 0, nil
	}
	targetEnd := rangesToQueue[len(rangesToQueue)-1].End
	var queuedUntil *int64
	if session.QueuedUntil.Valid && rangesToQueue[0].Start <= session.QueuedUntil.Int64 && session.QueuedUntil.Int64 < targetEnd {
		value := session.QueuedUntil.Int64
		queuedUntil = &value
		rangesToQueue = PlanMiddleRanges(session.MediaSize, headSize, tailSize, session.LastRangeEnd, queuedUntil, prefetch, middleCache)
	}
	inserted := 0
	var highestEnd *int64
	for _, byteRange := range rangesToQueue {
		if byteRange.End > targetEnd {
			byteRange.End = targetEnd
		}
		existing, err := store.ReusablePrefetchRanges(session.CacheKey, byteRange)
		if err != nil {
			return inserted, err
		}
		missing := subtractRanges(byteRange, existing)
		if len(missing) == 0 {
			value := byteRange.End
			highestEnd = &value
			continue
		}
		for _, gap := range missing {
			task, err := store.EnqueuePrefetchTask(session.ItemID, session.MediaSourceID, session.CacheKey, gap.Start, gap.End, priority, now, prefetch.MaxQueueDepth)
			if err != nil {
				return inserted, err
			}
			if task == nil {
				exists, err := store.PrefetchTaskExists(session.CacheKey, gap.Start, gap.End)
				if err != nil {
					return inserted, err
				}
				if exists {
					value := gap.End
					highestEnd = &value
					continue
				}
				if highestEnd != nil {
					return inserted, store.UpdateSessionQueuedUntil(session.SessionHash, *highestEnd, now)
				}
				return inserted, nil
			}
			inserted++
			value := gap.End
			highestEnd = &value
		}
	}
	if highestEnd != nil {
		if err := store.UpdateSessionQueuedUntil(session.SessionHash, *highestEnd, now); err != nil {
			return inserted, err
		}
	}
	return inserted, nil
}

func subtractRanges(byteRange model.ByteRange, existing []model.ByteRange) []model.ByteRange {
	var missing []model.ByteRange
	current := byteRange.Start
	for _, ex := range existing {
		if ex.End < current {
			continue
		}
		if ex.Start > byteRange.End {
			break
		}
		if ex.Start > current {
			end := ex.Start - 1
			if end > byteRange.End {
				end = byteRange.End
			}
			missing = append(missing, model.ByteRange{Start: current, End: end})
		}
		if ex.End+1 > current {
			current = ex.End + 1
		}
		if current > byteRange.End {
			break
		}
	}
	if current <= byteRange.End {
		missing = append(missing, model.ByteRange{Start: current, End: byteRange.End})
	}
	return missing
}

func alignDown(value, alignment int64) int64 {
	return value - value%alignment
}

func alignUp(value, alignment int64) int64 {
	if value%alignment == 0 {
		return value
	}
	return value + alignment - value%alignment
}

func minPositive(values ...int64) int64 {
	min := int64(0)
	for _, value := range values {
		if value <= 0 {
			return 0
		}
		if min == 0 || value < min {
			min = value
		}
	}
	return min
}

type BandwidthLimiter struct {
	bytesPerSecond int64
}

func NewBandwidthLimiter(bytesPerSecond int64) *BandwidthLimiter {
	return &BandwidthLimiter{bytesPerSecond: bytesPerSecond}
}

func (l *BandwidthLimiter) Consume(byteCount int) {
	if l == nil || l.bytesPerSecond <= 0 || byteCount <= 0 {
		return
	}
	time.Sleep(time.Duration(int64(byteCount) * int64(time.Second) / l.bytesPerSecond))
}

type limitedReader struct {
	reader  io.Reader
	limiter *BandwidthLimiter
}

func (r *limitedReader) Read(p []byte) (int, error) {
	n, err := r.reader.Read(p)
	r.limiter.Consume(n)
	return n, err
}

type originLimiter struct {
	limit int
	mu    sync.Mutex
	sems  map[string]chan struct{}
}

func newOriginLimiter(limit int) *originLimiter {
	if limit <= 0 {
		limit = 1
	}
	return &originLimiter{limit: limit, sems: make(map[string]chan struct{})}
}

func (l *originLimiter) acquire(rawURL string) func() {
	host := rawURL
	if parsed, err := url.Parse(rawURL); err == nil && parsed.Host != "" {
		host = parsed.Host
	}
	l.mu.Lock()
	sem := l.sems[host]
	if sem == nil {
		sem = make(chan struct{}, l.limit)
		l.sems[host] = sem
	}
	l.mu.Unlock()
	sem <- struct{}{}
	return func() { <-sem }
}

func ShortHash(value string) string {
	if value == "" {
		return "none"
	}
	if len(value) < 12 {
		return value
	}
	return value[:12]
}

var ErrSourceMismatch = fmt.Errorf("prefetch source metadata mismatch")
